src/logics/Logic.cc

```diff
@@ -1263,14 +1263,14 @@ lbool Logic::retrieveSubstitutions(const vec<PtAsgn>& facts, Map<PTRef,PtAsgn,PT
     return l_Undef;
 }
 
-void Logic::substitutionsTransitiveClosure(Map<PTRef, PtAsgn, PTRefHash> & substs) {
+void Logic::substitutionsTransitiveClosure(MapWithKeys<PTRef, PtAsgn, PTRefHash> & substs) {
     bool changed = true;
-    auto keyValPairs = substs.getKeysAndValsPtrs(); // We can use direct pointers, since no elements are inserted or deleted in the loop
-    std::vector<char> notChangedElems(keyValPairs.size(), 0); // True if not changed in last iteration, initially False
+    auto & keys(substs.getKeys()); // We can use direct pointers, since no elements are inserted or deleted in the loop
+    std::vector<char> notChangedElems(substs.getSize(), 0); // True if not changed in last iteration, initially False
     while (changed) {
         changed = false;
-        for (int i = 0; i < keyValPairs.size(); ++i) {
-            auto & val = keyValPairs[i]->data;
+        for (int i = 0; i < keys.size(); ++i) {
+            auto & val = substs[keys[i]];
             if (val.sgn != l_True || notChangedElems[i]) { continue; }
             PTRef oldVal = val.tr;
             PTRef newVal = Substitutor(*this, substs).rewrite(oldVal);
```

Comment:
A side note for future: since `Substitutor` depends on Logic, I don't think it should be used inside Logic. I think this whole method `substitutionsTransitiveClosure` should be moved outside the `Logic` class and put somewhere among helper methods.

Reply:
I agree.  We should find it a better home in the future.